### ISSUE TITLE
Port collated batch view from DataLoaders.jl

### ISSUE THREAD

**Issue by lorenzoh** [JuliaML/MLUtils.jl#29]
This is the first part of porting functionality from DataLoaders.jl (ref #22).

This includes porting

- `BatchViewCollated` from [`batchview.jl`](https://github.com/lorenzoh/DataLoaders.jl/blob/master/src/batchview.jl)
  - Notes: the `BatchDim` machinery can probably be dropped if we can assume last dim = batch dim
-  `collate` from [`https://github.com/lorenzoh/DataLoaders.jl/blob/master/src/collate.jl`]

`batchview.jl` also includes an unimported `batchsize` helper. Do we want to have a proper `batchsize`?

**Comment by CarloLucibello** [comment]
I think `collate` could be entirely replaced by batch (one of the reasons that motivated me in doing #27)

**Comment by CarloLucibello** [comment]
Also, do we need something like collate? Why don't we just call `getobs(data, idxs)`?

**Comment by lorenzoh** [comment]
Sometimes you want a batch view that gives you a vector of observations (e.g. different-sized images), and sometimes you'll want to have them as a single array (or recursively collate observations in tuples/dicts...). For this, having something like `BatchViewCollated` is essential since it allows buffered loading of observations with `getobs!`.

**Comment by CarloLucibello** [comment]
@lorenzoh is this still necessary?

**Comment by lorenzoh** [comment]
Yup 👍 
This is necessary for porting the buffered batch view from DataLoaders.jl

**Comment by CarloLucibello** [comment]
I think that we already support any sensible combination of `DataLoader(; buffer, batchsize, collate)`.
